Ast/src/Lexer.cpp

```diff
@@ -535,37 +587,67 @@ Lexeme Lexer::readQuotedString()
             return Lexeme(Location(start, position()), Lexeme::BrokenString);
 
         case '\\':
+            readBackslashInString();
+            break;
+
+        default:
             consume();
-            switch (peekch())
-            {
-            case '\r':
-                consume();
-                if (peekch() == '\n')
-                    consume();
-                break;
+        }
+    }
 
-            case 0:
-                break;
+    consume();
 
-            case 'z':
-                consume();
-                while (isSpace(peekch()))
-                    consume();
-                break;
+    return Lexeme(Location(start, position()), Lexeme::QuotedString, &buffer[startOffset], offset - startOffset - 1);
+}
 
-            default:
-                consume();
-            }
+Lexeme Lexer::readInterpolatedStringBegin()
+{
+    LUAU_ASSERT(peekch() == '`');
+
+    Position start = position();
+    consume();
+
+    return readInterpolatedStringSection(start, Lexeme::InterpStringBegin, Lexeme::InterpStringSimple);
+}
+
+Lexeme Lexer::readInterpolatedStringSection(Position start, Lexeme::Type formatType, Lexeme::Type endType)
+{
+    unsigned int startOffset = offset;
+
+    while (peekch() != '`')
+    {
+        switch (peekch())
+        {
+        case 0:
+        case '\r':
+        case '\n':
+            // INTERP TODO: Clear anything we've added to the brace stack, and write a test to see what happens if we don't
+            return Lexeme(Location(start, position()), Lexeme::BrokenString);
+
+        case '\\':
+            readBackslashInString();
             break;
 
+        case '{':
+        {
+            if (peekch(1) == '{')
+            {
+                return Lexeme(Location(start, position()), Lexeme::BrokenInterpDoubleBrace);
+            }
+
+            braceStack.push_back(BraceType::InterpolatedString);
```

Comment:
I think we want to push this before we throw `BrokenInterpDoubleBrace`? There's a test for `` `{{oops}}` ``, but it only observes the output of the parser errors, _not_ the lexeme types. That means this:

```
`foo{{bad}}bar`
```
is lexed something along the line of
```
BrokenInterpDoubleBrace(`foo{{`)
Ident(bad)
ClosingBrace
ClosingBrace
Ident(bar)
BrokenString(`)
```
which is not super great from the highlighting point of view, we want it to be
```
BrokenInterpDoubleBrace(`foo{{`)
Ident(bad)
InterpStringEnd(}}baz`)
```

Reply:
The output looks like this:

```cpp
    const std::string testInput = R"(`foo{{bad}}bar`)";
    Luau::Allocator alloc;
    AstNameTable table(alloc);
    Lexer lexer(testInput.c_str(), testInput.size(), table);

    auto brokenInterpBegin = lexer.next();
    CHECK_EQ(brokenInterpBegin.type, Lexeme::BrokenInterpDoubleBrace);
    CHECK_EQ(std::string(brokenInterpBegin.data, brokenInterpBegin.length), std::string("foo"));

    CHECK_EQ(lexer.next().type, Lexeme::Name);

    auto interpEnd = lexer.next();
    CHECK_EQ(interpEnd.type, Lexeme::InterpStringEnd);
    CHECK_EQ(std::string(interpEnd.data, interpEnd.length), std::string("}bar"));

```

Hope that is satisfactory

Reply:
Yup, that's what I wanted to see. Thanks!